networkx/drawing/nx_pylab.py

```diff
@@ -890,6 +965,95 @@ def _connectionstyle(posA, posB, *args, **kwargs):
     return edge_viz_obj
 
 
+def _draw_networkx_multigraph_edges(
+    G,
+    pos,
+    edgelist=None,
+    width=1.0,
+    edge_color="k",
+    style="solid",
+    alpha=None,
+    arrowstyle=None,
+    arrowsize=10,
+    edge_cmap=None,
+    edge_vmin=None,
+    edge_vmax=None,
+    ax=None,
+    label=None,
+    node_size=300,
+    nodelist=None,
+    node_shape="o",
+    connectionstyle="arc3",
+    min_source_margin=0,
+    min_target_margin=0,
+):
+    from collections import Counter
+
+    if edgelist == None:
+        edgelist = list(G.edges())
+
+    edge_counts = Counter(edgelist)
+
+    drawnList = []
+
+    for edge, cnt in edge_counts.items():
+        if cnt == 1 and list(G.edges()).count((edge[1], edge[0])) == 0:
+            drawnEdge = _draw_networkx_edges(
+                G,
+                pos,
+                edgelist=[edge],
+                width=width,
+                edge_color=edge_color,
+                style=style,
+                alpha=alpha,
+                arrowstyle=arrowstyle,
+                arrowsize=arrowsize,
+                edge_cmap=edge_cmap,
+                edge_vmin=edge_vmin,
+                edge_vmax=edge_vmax,
+                ax=ax,
+                arrows=True,
+                label=label,
+                node_size=node_size,
+                nodelist=nodelist,
+                node_shape=node_shape,
+                connectionstyle="arc3,rad=0",
+                min_source_margin=min_source_margin,
+                min_target_margin=min_target_margin,
+            )
+            drawnList.extend(drawnEdge)
+        else:
+            bend = 0.5 / cnt
+            rad = bend
+            for i in range(cnt):
+                drawnEdge = _draw_networkx_edges(
+                    G,
+                    pos,
+                    edgelist=[edge],
+                    width=width,
+                    edge_color=edge_color,
+                    style=style,
+                    alpha=alpha,
+                    arrowstyle=arrowstyle,
+                    arrowsize=arrowsize,
+                    edge_cmap=edge_cmap,
```

Comment:
Yeaah, scratch my previous comment. It was the wrong reasoning.

The problem here is that this logic plots one `edge` at a time while the `_draw_networkx_edges` logic for extracting `edge_color` from `edge_cmap` is written under the assumption that all the edges are present while plotting it.

https://github.com/networkx/networkx/blob/main/networkx/drawing/nx_pylab.py#L682-L700 extracts the color mapping from a `cmap` to `edge_color` but this fails as there is only one `edge` being passed here in the `edgelist`. Ideally the `edge_color` should already be extracted here from the `edge_cmap` here before passing it to the `_draw_networkx_edges` to plot that one edge.


I can't think of a not-too-invasive of fixing this right now :/

Reply:
Nice!  
I see that it is a tricky issue though.  When `edge_color` is an array, we will need to track the colors assigned to each multiedge and loop over them along with the edge itself. I guess if there are other arguments that allow arrays which match the order of `edgelist` we will need to do the same for them.

The basic idea is the same as what is in the PR: call `_draw_networkx_edges` once for each edge with `edgelist=[edge]`, only now we also need to pull out the colors from `edge_colors` as we loop over `edgelist` and use input arguments `edge_color=[ecolor]`.  I think that also means when multiedges are present that we have to collect not just the counts of how many multiedges there are, but also the colors of those edges.

The collective looping should be done for arguments `width`, `style` and `edge_color` with values collected along with edgelist. Perhaps we can construct a list for each edge with values `edge`, `width`, `style`, `edge_color`, `rad`. Then loop over that container of lists calling the `_draw_networkx_edges` function for each one.
`_draw_networkx_edges(G, pos, edgelist=[edge], width=[ewidth], edge_color=[ecolor], style=[estyle], ...)`.  The hard part I think is computing `rad` because you have to know how many multiedges there are before you assign those values and the values are ordered by edgelist, not necessarily with all the multiedges together.  You could collect the `edge_counts` first and then keep track of how many times each edge has appeared as you construct this container of edge information.

```python
num_edges = len(edgelist)
edge_counts = Counter(edgelist)
edge_so_far = dict.fromkeys(edge_counts, 0)  # make dict keyed by edge to value 0

widthlist = width if hasattr(width, "__iter__") else [width] * num_edges 
colorlist = edge_color if hasattr(edge_color, "__iter__") else [edge_color] * num_edges 
stylelist = style if hasattr(style, "__iter__") else [style] * num_edges 
all_edge_info = list(zip(edgelist, widthlist, colorlist, stylelist)) 
for edge, ewidth, ecolor, estyle in all_edge_info:
    rad = edge_so_far[edge] * (0.5 / edge_counts[edge])
    drawnEdge = _draw_networkx_edges(G, pos, edgelist=[edge], width=ewidth, edge_color=ecolor, style=estyle, ..., connectionstyle="arc3,rad=" + str(rad)[:3], ...)
    edge_so_far[edge] += 1
```
(probably a lot of stuff left out of here... you can come up with something better perhaps.)